File: backend/src/main/java/com/notitas/api/controller/AuthController.java
package com.notitas.api.controller;

import com.notitas.api.model.User;
import com.notitas.api.payload.ForgotPasswordRequest;
import com.notitas.api.payload.JwtResponse;
import com.notitas.api.payload.LoginRequest;
import com.notitas.api.payload.MessageResponse;
import com.notitas.api.payload.RegisterRequest;
import com.notitas.api.payload.ResetPasswordRequest;
import com.notitas.api.repository.UserRepository;
import com.notitas.api.security.JwtUtils;
import com.notitas.api.security.UserDetailsImpl;
import com.notitas.api.service.EmailService;
import com.notitas.api.service.PasswordResetService;
import io.jsonwebtoken.Claims;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    @Autowired
    AuthenticationManager authenticationManager;

    @Autowired
    UserRepository userRepository;

    @Autowired
    PasswordEncoder encoder;

    @Autowired
    JwtUtils jwtUtils;

    @Autowired
    PasswordResetService passwordResetService;

    @Autowired
    EmailService emailService;

    /**
     * Solo fuera de producción (dev/tests) se expone el enlace de reset en la
     * respuesta cuando no hay email configurado, para poder probar el flujo
     * localmente sin enviar correos.
     */
    @Value("${app.email.expose-reset-link:false}")
    private boolean exposeResetLink;

    @Value("${app.cookie.secure:false}")
    private boolean cookieSecure;

    /** SameSite de la cookie JWT (None en prod cross-site; Lax/Strict si la
     *  API y el frontend comparten dominio). Configurable por env var. */
    @Value("${app.cookie.samesite:None}")
    private String cookieSameSite;

    /** Cookie de sesión: se borra al cerrar el navegador. */
    private static final int SESSION_COOKIE_MAX_AGE = -1;

    /**
     * Expiración de la cookie "recordarme", en ms. La MISMA propiedad que usa
     * JwtUtils para el token, así cookie y JWT nunca se desincronizan. Se
     * convierte a segundos al fijar el maxAge de la cookie.
     */
    @Value("${app.jwt.remember-me-expiration-ms}")
    private long rememberMeExpirationMs;

    @PostMapping("/login")
    public ResponseEntity<?> authenticateUser(@Valid @RequestBody LoginRequest loginRequest,
                                              HttpServletResponse response) {
        try {
            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            loginRequest.getEmail(), loginRequest.getPassword()));

            boolean rememberMe = loginRequest.isRememberMe();
            String jwt = jwtUtils.generateJwtToken(authentication, rememberMe);

            UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();

            addJwtCookie(response, jwt, rememberMe);

            return ResponseEntity.ok(new JwtResponse(jwt, userDetails.getId(),
                    userDetails.getEmail(), userDetails.getName(),
                    userDetails.getAvatar()));
        } catch (AuthenticationException e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new MessageResponse("Credenciales inválidas"));
        }
    }

    @PostMapping("/register")
    public ResponseEntity<?> registerUser(@Valid @RequestBody RegisterRequest registerRequest) {
        if (userRepository.existsByEmail(registerRequest.getEmail())) {
            return ResponseEntity.badRequest()
                    .body(new MessageResponse("Error: El email ya está registrado"));
        }

        User user = new User(
                registerRequest.getName(),
                registerRequest.getEmail(),
                encoder.encode(registerRequest.getPassword())
        );
        userRepository.save(user);

        return ResponseEntity.ok(new MessageResponse("Usuario registrado exitosamente"));
    }

    /**
     * Logout con revocación real: incrementa el token_version del usuario para
     * invalidar todos los JWT emitidos antes (esta sesión y cualquier otra
     * copia del token), y borra la cookie. Funciona también con tokens ya
     * expirados (solo se necesita una firma válida para identificar al usuario).
     */
    /**
     * Recuperación de contraseña (paso 1): genera un token de un solo uso y
     * envía el enlace por email. La respuesta es idéntica exista o no la
     * cuenta, para no revelar qué emails están registrados.
     */
    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        String resetLink = passwordResetService.createPasswordResetLink(request.getEmail());

        Map<String, Object> body = new HashMap<>();
        body.put("message", "Si el correo está registrado, recibirás un enlace para restablecer tu contraseña.");
        // En dev/tests sin email configurado se devuelve el enlace para probar
        // el flujo completo sin depender de un proveedor de correo.
        if (resetLink != null && exposeResetLink && !emailService.isConfigured()) {
            body.put("devResetLink", resetLink);
        }
        return ResponseEntity.ok(body);
    }

    /**
     * Recuperación de contraseña (paso 2): valida el token y fija la nueva
     * contraseña. Revoca todas las sesiones activas del usuario.
     */
    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@Valid @RequestBody ResetPasswordRequest request) {
        boolean ok = passwordResetService.resetPassword(request.getToken(), request.getPassword());
        if (!ok) {
            return ResponseEntity.badRequest()
                    .body(new MessageResponse("El enlace de recuperación es inválido o ha expirado"));
        }
        return ResponseEntity.ok(new MessageResponse("Contraseña actualizada. Ya puedes iniciar sesión con tu nueva contraseña."));
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logoutUser(HttpServletRequest request, HttpServletResponse response) {
        String jwt = JwtUtils.extractJwt(request);
        if (jwt != null) {
            Claims claims = jwtUtils.parseClaimsIgnoreExpiration(jwt);
            if (claims != null) {
                userRepository.findByEmail(claims.getSubject()).ifPresent(user -> {
                    user.setTokenVersion(user.getTokenVersion() + 1);
                    userRepository.save(user);
                });
            }
        }

        Cookie cookie = new Cookie("jwt", "");
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        cookie.setMaxAge(0);
        cookie.setSecure(cookieSecure);
        cookie.setAttribute("SameSite", cookieSameSite);
        response.addCookie(cookie);

        return ResponseEntity.ok(new MessageResponse("Sesión cerrada exitosamente"));
    }

    /**
     * Renovación deslizante: si la cookie JWT es válida (firma + no expirado +
     * versión de sesión vigente), emite un JWT nuevo y renueva la cookie.
     * El frontend lo llama al arrancar y periódicamente, así la sesión no se
     * corta a las 24 h mientras la app está en uso. Los tokens expirados o
     * revocados reciben 401 (el frontend cierra sesión limpiamente).
     */
    @PostMapping("/refresh")
    public ResponseEntity<?> refreshSession(HttpServletRequest request, HttpServletResponse response) {
        String jwt = JwtUtils.extractJwt(request);
        if (jwt == null || !jwtUtils.validateJwtToken(jwt)) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new MessageResponse("Sesión no válida o expirada"));
        }

        String email = jwtUtils.getUserNameFromJwtToken(jwt);
        User user = userRepository.findByEmail(email).orElse(null);
        if (user == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new MessageResponse("Usuario no encontrado"));
        }
        if (jwtUtils.getTokenVersionFromJwtToken(jwt) != user.getTokenVersion()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new MessageResponse("Sesión revocada"));
        }

        // El refresh conserva la marca "recordarme" del token: ni degrada una
        // sesión persistente a cookie de navegador, ni promociona una sesión
        // normal a persistente.
        boolean rememberMe = jwtUtils.getRememberMeFromJwtToken(jwt);
        String newToken = jwtUtils.generateTokenFromUsername(user.getEmail(), user.getTokenVersion(), rememberMe);
        addJwtCookie(response, newToken, rememberMe);
        return ResponseEntity.ok(new JwtResponse(newToken, user.getId(), user.getEmail(), user.getName(), user.getAvatar()));
    }

    private void addJwtCookie(HttpServletResponse response, String jwt, boolean rememberMe) {
        Cookie cookie = new Cookie("jwt", jwt);
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        // rememberMe → 30 días (casado con la expiración del JWT); si no,
        // cookie de sesión que se borra al cerrar el navegador.
        cookie.setMaxAge(rememberMe ? (int) (rememberMeExpirationMs / 1000) : SESSION_COOKIE_MAX_AGE);
        cookie.setSecure(cookieSecure);
        cookie.setAttribute("SameSite", cookieSameSite);
        response.addCookie(cookie);
    }
}

File: backend/src/controllers/authController.js
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import crypto from 'crypto';
import { query } from '../config/db.js';
import { sendPasswordResetEmail } from '../services/emailService.js';
import dotenv from 'dotenv';

dotenv.config();

const JWT_SECRET = process.env.NOTITAS_JWT_SECRET || process.env.JWT_SECRET || 'notitas-super-secret-jwt-key-2026-production';
const COOKIE_SECURE = process.env.NODE_ENV === 'production' || process.env.COOKIE_SECURE === 'true';
const COOKIE_SAMESITE = process.env.COOKIE_SAMESITE || (process.env.NODE_ENV === 'production' ? 'none' : 'lax');

const generateToken = (user, rememberMe = false) => {
  const expiresIn = rememberMe ? '30d' : '24h';
  return jwt.sign(
    {
      id: user.id,
      sub: user.email,
      email: user.email,
      tv: Number(user.token_version || 0),
      rm: Boolean(rememberMe),
    },
    JWT_SECRET,
    { expiresIn }
  );
};

const setJwtCookie = (res, token, rememberMe = false) => {
  const maxAge = rememberMe ? 30 * 24 * 60 * 60 * 1000 : undefined;
  res.cookie('jwt', token, {
    httpOnly: true,
    secure: COOKIE_SECURE,
    sameSite: COOKIE_SAMESITE,
    path: '/',
    maxAge,
  });
};

export const register = async (req, res, next) => {
  try {
    const { name, email, password } = req.body;

    if (!name || !email || !password) {
      return res.status(400).json({ message: 'Todos los campos son obligatorios' });
    }

    const emailNorm = email.toLowerCase().trim();
    const existing = await query('SELECT id FROM users WHERE LOWER(email) = $1', [emailNorm]);

    if (existing.rows.length > 0) {
      return res.status(400).json({ message: 'Error: El email ya está registrado' });
    }

    const hashedPassword = await bcrypt.hash(password, 10);
    await query(
      'INSERT INTO users (name, email, password, token_version) VALUES ($1, $2, $3, 0)',
      [name.trim(), emailNorm, hashedPassword]
    );

    return res.status(200).json({ message: 'Usuario registrado exitosamente' });
  } catch (error) {
    next(error);
  }
};

export const login = async (req, res, next) => {
  try {
    const { email, password, rememberMe } = req.body;

    if (!email || !password) {
      return res.status(400).json({ message: 'Email y contraseña requeridos' });
    }

    const emailNorm = email.toLowerCase().trim();
    const result = await query('SELECT * FROM users WHERE LOWER(email) = $1', [emailNorm]);

    if (result.rows.length === 0) {
      return res.status(401).json({ message: 'Credenciales inválidas' });
    }

    const user = result.rows[0];
    const match = await bcrypt.compare(password, user.password);

    if (!match) {
      return res.status(401).json({ message: 'Credenciales inválidas' });
    }

    const token = generateToken(user, rememberMe);
    setJwtCookie(res, token, rememberMe);

    return res.status(200).json({
      token,
      id: Number(user.id),
      email: user.email,
      name: user.name,
      avatar: user.avatar,
    });
  } catch (error) {
    next(error);
  }
};

export const refresh = async (req, res, next) => {
  try {
    let token = req.cookies?.jwt || req.cookies?.token;
    if (!token && req.headers.authorization?.startsWith('Bearer ')) {
      token = req.headers.authorization.split(' ')[1];
    }

    if (!token) {
      return res.status(401).json({ message: 'Sesión no válida o expirada' });
    }

    let decoded;
    try {
      decoded = jwt.verify(token, JWT_SECRET);
    } catch {
      return res.status(401).json({ message: 'Sesión no válida o expirada' });
    }

    const result = await query('SELECT * FROM users WHERE id = $1', [decoded.id]);
    if (result.rows.length === 0) {
      return res.status(401).json({ message: 'Usuario no encontrado' });
    }

    const user = result.rows[0];
    if (decoded.tv !== undefined && Number(decoded.tv) !== Number(user.token_version)) {
      return res.status(401).json({ message: 'Sesión revocada' });
    }

    const rememberMe = Boolean(decoded.rm);
    const newToken = generateToken(user, rememberMe);
    setJwtCookie(res, newToken, rememberMe);

    return res.status(200).json({
      token: newToken,
      id: Number(user.id),
      email: user.email,
      name: user.name,
      avatar: user.avatar,
    });
  } catch (error) {
    next(error);
  }
};

export const logout = async (req, res, next) => {
  try {
    let token = req.cookies?.jwt || req.cookies?.token;
    if (!token && req.headers.authorization?.startsWith('Bearer ')) {
      token = req.headers.authorization.split(' ')[1];
    }

    if (token) {
      try {
        const decoded = jwt.decode(token);
        if (decoded && (decoded.id || decoded.sub)) {
          await query(
            'UPDATE users SET token_version = token_version + 1 WHERE id = $1 OR LOWER(email) = LOWER($2)',
            [decoded.id || 0, decoded.sub || '']
          );
        }
      } catch {
        // Ignorar error al decodificar token expirado
      }
    }

    res.clearCookie('jwt', {
      httpOnly: true,
      secure: COOKIE_SECURE,
      sameSite: COOKIE_SAMESITE,
      path: '/',
    });

    return res.status(200).json({ message: 'Sesión cerrada exitosamente' });
  } catch (error) {
    next(error);
  }
};

export const forgotPassword = async (req, res, next) => {
  try {
    const { email } = req.body;
    if (!email) {
      return res.status(400).json({ message: 'El correo electrónico es requerido' });
    }

    const emailNorm = email.toLowerCase().trim();
    const userRes = await query('SELECT id FROM users WHERE LOWER(email) = $1', [emailNorm]);

    let devResetLink = null;
    if (userRes.rows.length > 0) {
      const user = userRes.rows[0];
      const resetToken = crypto.randomBytes(32).toString('hex');
      const expiresAt = new Date(Date.now() + 60 * 60 * 1000); // 1 hora

      await query(
        'INSERT INTO password_reset_tokens (user_id, token, expires_at, used) VALUES ($1, $2, $3, false)',
        [user.id, resetToken, expiresAt]
      );

      const emailResult = await sendPasswordResetEmail(emailNorm, resetToken);
      if (emailResult.devLink) {
        devResetLink = emailResult.devLink;
      }
    }

    const responsePayload = {
      message: 'Si el correo está registrado, recibirás un enlace para restablecer tu contraseña.',
    };
    if (devResetLink && process.env.NODE_ENV !== 'production') {
      responsePayload.devResetLink = devResetLink;
    }

    return res.status(200).json(responsePayload);
  } catch (error) {
    next(error);
  }
};

export const resetPassword = async (req, res, next) => {
  try {
    const { token, password } = req.body;
    if (!token || !password) {
      return res.status(400).json({ message: 'Token y nueva contraseña son requeridos' });
    }

    const tokenRes = await query(
      'SELECT * FROM password_reset_tokens WHERE token = $1 AND used = false AND expires_at > NOW()',
      [token]
    );

    if (tokenRes.rows.length === 0) {
      return res.status(400).json({ message: 'El enlace de recuperación es inválido o ha expirado' });
    }

    const resetRecord = tokenRes.rows[0];
    const hashedPassword = await bcrypt.hash(password, 10);

    // Actualizar contraseña y revocar sesiones previas
    await query(
      'UPDATE users SET password = $1, token_version = token_version + 1 WHERE id = $2',
      [hashedPassword, resetRecord.user_id]
    );

    // Marcar token como usado
    await query('UPDATE password_reset_tokens SET used = true WHERE id = $1', [resetRecord.id]);

    return res.status(200).json({
      message: 'Contraseña actualizada. Ya puedes iniciar sesión con tu nueva contraseña.',
    });
  } catch (error) {
    next(error);
  }
};
